Fix scrolling

Dantemss opened the pull request:
- Tested (not extensively) for scrolling to:
  - Figures
  - Tables
  - Equations
  - Inline items
  - Solution links

- Made some tests for the id regex
- Made browser tests for scrolling to the right place

Dantemss committed:
Fixed scrolling to figures, tables, equations and inline terms

TomWoodward commented on src/app/content/components/Page.tsx:
```diff
@@ -58,6 +58,8 @@ export class PageComponent extends Component<PropTypes> {
       // same as previous, but allow indexing links to relative content
       .replace(/<a(.*?href="\.\.\/.*?)>/g, '<a target="_blank"$1>')
+      // move figure and table ids up to the parent div
+      .replace(/(<div[^>]*)(>(?!<\/?div>)*?<(?:figure|table)[^>]*?)( id=[^\s>]*)/g, '$1$3$2')
```
this works in chrome, but gives an error on https://regex101.com/ for the `*?`, substitution seems to work without it though

Dantemss replied:
I'll see if it accepts putting the `*?` inside the negative lookahead.

Dantemss replied:
Turns out... this expression would not match `<div><div></div><figure id="blah">` anyway... maybe I'll just simplify it to match the figure/table only if it is the first child of the div. We can worry about other cases when we stop doing this in regex.

TomWoodward replied:
i think they are always the first child anyway, so that should be fine

Dantemss committed:
Cleaned up id regex and CSS scrolling rules

Dantemss committed:
Fix equation numbers and solution links... when linked from the same page

TomWoodward committed:
Scrolling css (#206)

* use scrolling anchors

* different solutions for different element types

* use scroll-snap

TomWoodward commented on src/index.css:
```diff
@@ -15,6 +11,13 @@ body, #root, html {
+  font-size: 62.5%; /* Now 10px = 1rem! */
+  scroll-snap-type: y proximity;
+  scroll-padding-top: 120px;
+  overflow-y: scroll;
```
as we talked about in slack, we should try to move the `120px` (size of the nav) to a place that logically should have knowledge of that, and use the variables to get it.

also `overflow-y: scroll;` would ideally be `overflow: auto;`, hopefully that doesn't break everything 😬

Dantemss replied:
I think `overflow: auto` should work fine... I'll give it a try. And move the offset to the main content component as discussed.

Dantemss replied:
Seems I can put `overflow: auto` in the `MainContentWrapper` and it works fine. This is also right below the one element that really needs `overflow: visible` (the `div#root`)

Dantemss replied:
But... hopefully this doesn't break mobile *looks at the monster layout in Content.tsx*

Dantemss replied:
Ignore that last one, `overflow: auto` needs to be in the html tag and that seems to just work.

Dantemss committed:
Renamed scrollTo helper to scrollToContent
Keep track of navbar offset in scrollToContent

Dantemss committed:
Switched html element to overflow: auto

Dantemss committed:
Move scroll-padding-top to GlobalStyle in Content component

Dantemss committed:
Fixed failing tests (except a11y score)

Dantemss committed:
Fixed typing error and a11y scores

Dantemss committed:
Fix failing image screenshot tests

Dantemss committed:
Fixed lint errors

Dantemss committed:
Added unit test for id-moving regex and made it also set data-id

Dantemss committed:
Fixed scroll padding on mobile
Added snapshot tests for scrolling
Added more content to the generic styles page
Re-created 2 snapshots for the generic styles page

Dantemss committed:
Fix mobile JS scrolling

TomWoodward commented on src/app/content/components/Content.tsx:
```diff
@@ -148,29 +159,32 @@ const HideOverflowAndRedoPadding = isOpenConnector(styled.div`
-  </Background>
-</Layout>;
+const Content: React.SFC = () => <React.Fragment>
+  <GlobalStyle />
```
can't you put the globalstyle inside the layout and not have the `React.Fragment` thing?

Dantemss replied:
OK I'll do that.

TomWoodward commented on src/app/components/SkipToContentWrapper.tsx:
```diff
@@ -19,7 +20,8 @@ export default class SkipToContentWrapper extends Component {
       event.preventDefault();
-      scrollTo(this.mainContent);
+      scrollToContent(this.mainContent);
+      (this.mainContent as HTMLElement).focus();
```
you shouldn't need this `as HTMLElement`, its already a `HTMLDivElement`, if you're getting the wrong `HTMLDivElement` you might have to import it from `lib.dom`

Dantemss replied:
I think it was complaining that HTMLDivElement did not respond to focus... I'll try the import and see what happens.

Dantemss replied:
If I add this import here, I also have to add it to SkipToContent.ts but then the mergeRefs function gets a type error because this HTMLDivElement is not assignable to the non-OpenStax HTMLElement and I'm not sure how to fix that one...

TomWoodward commented on src/app/content/components/Content.screenshotspec.tsx:
```diff
@@ -0,0 +1,43 @@
+      it('scrolls correctly to all elements', async() => {
+        await navigate(page, TEST_PAGE_URL);
+        await finishRender(page);
+        expect(await page.screenshot()).toMatchImageSnapshot(SNAPSHOT_OPTIONS);
```
can you do these as browserspecs by measuring the page scrollTop after every scroll against a known or computed-good value? the screenshots are brittle and add a lot of bytes to the repo. i prefer to use the screenshots when something just can't be reasonably tested with math (like: `it('looks good')`)

Dantemss replied:
Alright, I'll change them to browserspecs.

TomWoodward commented on src/app/content/components/Content.tsx:
```diff
@@ -26,6 +27,16 @@ import { isOpenConnector, styleWhenSidebarClosed } from './utils/sidebar';
+// tslint:disable-next-line:variable-name
+const GlobalStyle = createGlobalStyle`
+  html {
+    scroll-padding-top: ${(bookBannerDesktopMiniHeight + toolbarDesktopHeight) * .625}rem;
```
` * .625` gives me anxiety, i see why it is necessary, can we extract to a constant or helper function that says what this is doing?

Dantemss replied:
I'll put it in constants.

Dantemss replied:
I guess I should ask... why is the `.625` necessary anyway?
Shouldn't rem in the `<html>` element just use its own font size?
This behavior seems... odd to me...
I'll test to see if `em` behaves better in the html element.

Dantemss replied:
Seems like using `em` instead of `rem` here removes the need for the multiplier.

TomWoodward commented on src/app/content/components/Page.spec.tsx:
```diff
@@ -382,7 +441,7 @@ describe('Page', () => {
   });
 
-  it('scrolls to selected content on load', () => {
+  it('does not scroll to selected content on load', () => {
```
```suggestion
  it('does not scroll to selected content on initial load', () => {
```

TomWoodward commented on src/app/content/components/utils/scrollToContent.ts:
```diff
@@ -0,0 +1,21 @@
+    // innerWidth works best on Chrome
+    // check other browsers at https://ryanve.com/lab/dimensions/
+    // default to desktopOffset if window is undefined (at least the target is always visible)
+    const offset = window && window.innerWidth <= pxPerRem * mobileBreak ? mobileOffset : desktopOffset;
```
this is what `window.matchMedia(theme.breakpoints.mobileQuery)` is for

but as i wrote above, i think it would be better to `window.getComputedStyle(document.documentElement).scrollPaddingTop`

Dantemss replied:
I will change it to use `getComputedStyle`. This should also fix the missing coverage for this statement.

TomWoodward commented on src/app/components/SkipToContentWrapper.tsx:
```diff
@@ -1,6 +1,7 @@
 import React, { Component } from 'react';
 import { FormattedMessage } from 'react-intl';
-import scrollTo from '../../helpers/scrollTo';
+import scrollToContent from '../content/components/utils/scrollToContent';
```
lost this comment on the first review:

the directory `content` is for book-content related stuff, this helper is an accessibility thing "focus the main content of the page". this should be decoupled from anything happening in the content directory because we'll use it on pages that are not in that module.

the `scrollToContent` helper should be in `app/utils`, and the logic should be decoupled from the knowledge of the navbars, maybe reading the computed `scroll-padding-top` off the html element would be more flexible?

Dantemss replied:
Yes, I agree that with the change to the helper to use `getComputedStyle` it can live outside the content dir.

Dantemss committed:
Update src/app/content/components/Page.spec.tsx

Co-Authored-By: Thomas Woodward <tom.g.woodward@gmail.com>

Dantemss committed:
Addressed PR requested changes

Dantemss committed:
Merge branch 'scrolling' of https://github.com/openstax/rex-web into scrolling

Dantemss committed:
Removed exports that are no longer needed and extra line break

Dantemss commented:
@TomWoodward: I made all requested changes except the `HTMLDivElement` one because as mentioned I am not sure what to do about `mergeRefs` failing with that change.

Dantemss committed:
Merge branch 'master' into scrolling

Dantemss committed:
Fix failing spec

Dantemss committed:
Fixed code coverage

Dantemss committed:
Merge branch 'master' into scrolling

Dantemss committed:
Improve assertions in app/utils

Dantemss committed:
Merge branch 'scrolling' of https://github.com/openstax/rex-web into scrolling

Dantemss committed:
Fixed failing utils.spec.ts

TomWoodward committed:
fix some types

Dantemss committed:
Use jest.spyOn() instead of Object.defineProperty()

Dantemss committed:
Merge branch 'scrolling' of https://github.com/openstax/rex-web into scrolling

TomWoodward committed:
wait for all images to load before scrolling on js navigation

TomWoodward committed:
clear promises

TomWoodward committed:
Merge remote-tracking branch 'origin/master' into scrolling

TomWoodward committed:
code cov

TomWoodward committed:
fix types
